cloudinit/cmd/tests/test_query.py

```diff
@@ -1,195 +1,267 @@
+
+    def test_handle_args_error_when_no_read_permission_instance_data(
+        self, caplog, tmpdir
+    ):
         """When instance_data file is unreadable, log an error."""
-        noread_fn = self.tmp_path('unreadable', dir=self.tmp)
-        write_file(noread_fn, 'thou shall not pass')
+        noread_fn = tmpdir.join('unreadable')
+        noread_fn.write('thou shall not pass')
         args = self.args(
-            debug=False, dump_all=True, format=None, instance_data=noread_fn,
+            debug=False, dump_all=True, format=None,
+            instance_data=noread_fn.strpath,
             list_keys=False, user_data='ud', vendor_data='vd', varname=None)
-        with mock.patch('sys.stderr', new_callable=StringIO) as m_stderr:
-            with mock.patch('cloudinit.cmd.query.util.load_file') as m_load:
-                m_load.side_effect = OSError(errno.EACCES, 'Not allowed')
-                self.assertEqual(1, query.handle_args('anyname', args))
-        self.assertIn(
-            "ERROR: No read permission on '%s'. Try sudo" % noread_fn,
-            self.logs.getvalue())
-        self.assertIn(
-            "ERROR: No read permission on '%s'. Try sudo" % noread_fn,
-            m_stderr.getvalue())
+        with mock.patch('cloudinit.cmd.query.util.load_file') as m_load:
+            m_load.side_effect = OSError(errno.EACCES, 'Not allowed')
+            assert 1 == query.handle_args('anyname', args)
+        msg = "No read permission on '%s'. Try sudo" % noread_fn
+        assert msg in caplog.text
 
-    def test_handle_args_defaults_instance_data(self):
+    def test_handle_args_defaults_instance_data(self, caplog, tmpdir):
         """When no instance_data argument, default to configured run_dir."""
         args = self.args(
             debug=False, dump_all=True, format=None, instance_data=None,
             list_keys=False, user_data=None, vendor_data=None, varname=None)
-        run_dir = self.tmp_path('run_dir', dir=self.tmp)
-        ensure_dir(run_dir)
-        paths = Paths({'run_dir': run_dir})
-        self.add_patch('cloudinit.cmd.query.read_cfg_paths', 'm_paths')
-        self.m_paths.return_value = paths
-        with mock.patch('sys.stderr', new_callable=StringIO) as m_stderr:
-            self.assertEqual(1, query.handle_args('anyname', args))
-        json_file = os.path.join(run_dir, INSTANCE_JSON_FILE)
-        self.assertIn(
-            'ERROR: Missing instance-data file: %s' % json_file,
-            self.logs.getvalue())
-        self.assertIn(
-            'ERROR: Missing instance-data file: %s' % json_file,
-            m_stderr.getvalue())
+        paths, run_dir, _, _ = self._setup_paths(tmpdir)
+        with mock.patch('cloudinit.cmd.query.read_cfg_paths') as m_paths:
+            m_paths.return_value = paths
+            assert 1 == query.handle_args('anyname', args)
+        json_file = run_dir.join(INSTANCE_JSON_FILE)
+        msg = 'Missing instance-data file: %s' % json_file.strpath
```

Comment:
If this test requested the `paths` fixture instead of the `tmpdir` fixture, I think it can be rewritten as:

```suggestion
        with mock.patch('cloudinit.cmd.query.read_cfg_paths') as m_paths:
            m_paths.return_value = paths
            assert 1 == query.handle_args('anyname', args)
        json_file = os.path.join(paths.run_dir, INSTANCE_JSON_FILE)
        msg = 'Missing instance-data file: %s' % json_file
```

(Roughly the same also applies to the following test.)